### PR TITLE
Rashuai/cuda top k

### PR DESCRIPTION
**Pull request by RandySheriffH**
**Description**: Describe your changes.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

Implement TopK opset11 cuda.

### CODE REVIEW

File: onnxruntime/core/providers/cuda/math/topk.cc [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  auto tensor_X = ctx->Input<Tensor>(0);
+  ORT_ENFORCE(nullptr != tensor_X);
+  auto rank = static_cast<int64_t>(tensor_X->Shape().NumDimensions());
+  auto axis = axis_ < 0 ? rank + axis_ : axis_;
+  ORT_ENFORCE(axis > -1 && axis < rank);
+
+  auto tensor_K = ctx->Input<Tensor>(1);
+  ORT_ENFORCE(nullptr != tensor_K);
+  int64_t K = 0;
+  cudaMemcpyAsync(&K, tensor_K->Data<int64_t>(), sizeof(int64_t), cudaMemcpyDeviceToHost);
```
Comment (yufenglee) [inline onnxruntime/core/providers/cuda/math/topk.cc]: You can call .OutputMemoryType<OrtMemTypeCPUOutput>(1) in kernel register to force tensor_K as a cpu output.

File: onnxruntime/core/providers/cuda/math/topk_impl.cu [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+#include "topk_impl.h"
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include "core/providers/cuda/cuda_common.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+template <typename T>
+__global__ void TopKKernel(const T* input_x, T* output_v, int64_t* output_i, const int64_t* elem_nums, size_t size, int64_t axis, int64_t K, int64_t largest, int64_t sorted, int64_t N, int64_t dimension) {
```
Comment (KeDengMS) [inline onnxruntime/core/providers/cuda/math/topk_impl.cu]: >TopKKernel [](start = 16, length = 10)

Is this faster than radix sort from CUB? [https://github.com/dmlc/cub/blob/master/cub/device/device_radix_sort.cuh](https://github.com/dmlc/cub/blob/master/cub/device/device_radix_sort.cuh)

File: onnxruntime/core/providers/cuda/math/topk.cc [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  if (0 == K) {
+    return Status::OK();
+  }
+
+  auto elem_nums = tensor_X->Shape().GetDims();
+  for (auto i = static_cast<int32_t>(elem_nums.size())-2; i >= 0; --i) {
+    elem_nums[i] *= elem_nums[i+1];
+  }
+
+  auto dimension = axis == elem_nums.size()-1 ? elem_nums[axis] : elem_nums[axis] / elem_nums[axis+1];
```
Comment (yufenglee) [inline onnxruntime/core/providers/cuda/math/topk.cc]: auto dimension = tensor_X->Shape()[axis] ?

File: onnxruntime/core/providers/cuda/math/topk_impl.cu [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    auto output_offset = left + (heap.Size() - 1) * (axis == size - 1 ? 1 : elem_nums[axis + 1]) + right;
+    output_v[output_offset] = node.value_;
+    output_i[output_offset] = node.index_;
+    heap.Pop();
+  }
+}
+
+template <typename T>
+Status TopKImpl(const T* input_x, T* output_v, int64_t* output_i, const int64_t* elem_nums, size_t size, int64_t axis, int64_t K, int64_t largest, int64_t sorted, int64_t N, int64_t dimension) {
+  int blocksPerGrid = (int)(ceil(static_cast<float>(N) / GridDim::maxThreadsPerBlock));
```
Comment (yufenglee) [inline onnxruntime/core/providers/cuda/math/topk_impl.cu]: so you are implementing a TopK in one single thread. So for a tensor with shape [1, 4, 4096], k=5 and axis = 2, there is only 4 thread. 
Have you test the performance?

File: onnxruntime/core/providers/cuda/math/topk_impl.cu (hunk 2 above)
Comment (yufenglee) [reply onnxruntime/core/providers/cuda/math/topk_impl.cu]: Those may help too:
[https://anilshanbhag.in/static/papers/gputopk_sigmod18.pdf](https://anilshanbhag.in/static/papers/gputopk_sigmod18.pdf)
[https://github.com/anilshanbhag/gpu-topk](https://github.com/anilshanbhag/gpu-topk) 
[https://github.com/tensorflow/tensorflow/issues/5719](https://github.com/tensorflow/tensorflow/issues/5719)

---
In reply to: [328254646](https://github.com/microsoft/onnxruntime/pull/1919#discussion_r328254646) [](ancestors = 328254646)

Comment (tianleiwu) [reply onnxruntime/core/providers/cuda/math/topk_impl.cu]: radix selection algorithm seems to be a good choice. I saw another one besides gpu-topk: https://github.com/facebook/fbcuda/blob/master/TopK.cuh

Comment (ybrnathan) [reply onnxruntime/core/providers/cuda/math/topk_impl.cu]: It seems like both PyTorch and TF are using CUB, Cub is required for NomMaxSuppression op as well, we might want to include it as dependency.

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 5]
```diff
@@ -538,6 +538,8 @@ class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain,
 
 class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 10, Dropout);
 
+class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 11, TopK);
```
Comment (hariharans29) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc]: Hi Randy - Can we also register kernels for TopK opset-9 and opset-10 along with this change ? It seems like by supporting opset-11 TopK, we are inherently supporting the previous opsets too (which have sorted ==1 and largest == 1 always). So, models stamped with opset-9 or opset-10 can get the perf gain too - (#1830)

File: onnxruntime/core/providers/cuda/math/topk_impl.cu (hunk 2 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/cuda/math/topk_impl.cu]: Using Cub now.

File: onnxruntime/core/providers/cuda/math/topk.cc (hunk 1 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/cuda/math/topk.cc]: Line removed.

File: onnxruntime/core/providers/cuda/math/topk.cc (hunk 3 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/cuda/math/topk.cc]: line removed.

File: onnxruntime/core/providers/cuda/math/topk_impl.cu (hunk 4 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/cuda/math/topk_impl.cu]: Using cub devcie sort now.

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc (hunk 5 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/cuda/cuda_execution_provider.cc]: Like this?

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 6]
```diff
@@ -536,17 +536,21 @@ class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain,
 class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 9, float, Less);
 class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 9, double, Less);
 class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 9, MLFloat16, Less);
+class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 9, TopK);
```
Comment (ybrnathan) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc]: versioned kernel 1,9

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 7]
```diff
@@ -536,17 +536,21 @@ class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain,
...
 
 // opset 10
 class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 10, Dropout);
 class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 10, float, RoiAlign);
 class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 10, double, RoiAlign);
+class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 10, TopK);
```
Comment (ybrnathan) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc]: versioned kernel 10, 10

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 8]
```diff
@@ -891,6 +895,7 @@ static void RegisterCudaKernels(KernelRegistry& kernel_registry) {
       BuildKernelCreateInfo<ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 11, MLFloat16, Gemm)>,
       BuildKernelCreateInfo<ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 11, float, Gemm)>,
       BuildKernelCreateInfo<ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 11, double, Gemm)>,
+      BuildKernelCreateInfo<ONNX_OPERATOR_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 11, TopK)>,
```
Comment (ybrnathan) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc]: add versioned build kernel info for (1,9), (10, 10)

File: onnxruntime/core/providers/cuda/math/topk.cc [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+#include "topk.h"
+#include "topk_impl.h"
+// #include "core/providers/cuda/cuda_allocator.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+ONNX_OPERATOR_KERNEL_EX(
+  TopK,
```
Comment (ybrnathan) [inline onnxruntime/core/providers/cuda/math/topk.cc]: versioned 1, 9

File: onnxruntime/core/providers/cuda/math/topk.cc [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+  kOnnxDomain,
+  9,
+  kCudaExecutionProvider,
+  KernelDefBuilder().TypeConstraint("T", DataTypeImpl::AllFixedSizeTensorTypes()),
+  TopK);
+
+ONNX_OPERATOR_KERNEL_EX(
+  TopK,
```
Comment (ybrnathan) [inline onnxruntime/core/providers/cuda/math/topk.cc]: versioned 10, 10

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc (hunk 6 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/cuda/cuda_execution_provider.cc]: done.

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc (hunk 8 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/cuda/cuda_execution_provider.cc]: done.

File: onnxruntime/core/providers/cuda/math/topk.cc (hunk 10 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/cuda/math/topk.cc]: done.

File: onnxruntime/core/providers/cuda/math/topk.cc [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+  auto tensor_X = ctx->Input<Tensor>(0);
+  ORT_ENFORCE(nullptr != tensor_X);
+  auto rank = static_cast<int64_t>(tensor_X->Shape().NumDimensions());
+  auto axis = axis_ < 0 ? rank + axis_ : axis_;
+  ORT_ENFORCE(axis > -1 && axis < rank);
+
+  if (inputk) {
+    auto tensor_K = ctx->Input<Tensor>(1);
+    ORT_ENFORCE(nullptr != tensor_K);
+    cudaMemcpyAsync(&K_, tensor_K->Data<int64_t>(), sizeof(int64_t), cudaMemcpyDeviceToHost);
```
Comment (yufenglee) [inline onnxruntime/core/providers/cuda/math/topk.cc]: >cudaMemcpyAsync(&K_, tensor_K->Data<int64_t>(), sizeof(int64_t), cudaMemcpyDeviceToHost); [](start = 4, length = 89)

cudaMemcpyAsync is async with CPU. It will return immediately.

Comment (RandySheriffH) [reply onnxruntime/core/providers/cuda/math/topk.cc]: done.

File: onnxruntime/core/providers/cuda/math/topk.cc [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+  if (0 == K_) {
+    return Status::OK();
+  }
+
+  auto elem_nums = tensor_X->Shape().GetDims();
+  for (auto i = static_cast<int32_t>(elem_nums.size()) - 2; i >= 0; --i) {
+    elem_nums[i] *= elem_nums[i + 1];
+  }
+
+  auto dimension = axis == elem_nums.size() - 1 ? elem_nums[axis] : elem_nums[axis] / elem_nums[axis + 1];
```
Comment (yufenglee) [inline onnxruntime/core/providers/cuda/math/topk.cc]: nit: move this ahead of the for loop and you can change it to "auto dimension = elem_nums[axis]; "

File: onnxruntime/core/providers/cuda/math/topk_impl.cu [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+template <typename T>
+Status TopKImpl(const T* input_x, T* output_v, int64_t* output_i, const int64_t* elem_nums, size_t size, int64_t axis, int64_t K, int64_t largest, int64_t sorted, int64_t N, int64_t dimension) {
+  auto err = cudaSuccess;
+  T* input_key;
+  T* output_key;
+  int64_t* input_value;
+  int64_t* output_value;
+  err = cudaMalloc(&input_key, sizeof(T) * dimension);
```
Comment (yufenglee) [inline onnxruntime/core/providers/cuda/math/topk_impl.cu]: >cudaMalloc [](start = 8, length = 10)

Please use CudaAsyncBuffer. cudaMalloc is a sync call. Will block the whole computation.

File: onnxruntime/core/providers/cuda/math/topk_impl.cu [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+  err = cudaSuccess == err ? cudaMalloc(&output_key, sizeof(T) * dimension) : err;
```
Comment (KeDengMS) [inline onnxruntime/core/providers/cuda/math/topk_impl.cu]: >err = cudaSuccess == err ?  [](start = 2, length = 27)

Preferred CUDA_RETURN_IF_ERROR for early return when error

File: onnxruntime/core/providers/cuda/math/topk.cc (hunk 12 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/cuda/math/topk.cc]: Done.

File: onnxruntime/core/providers/cuda/math/topk_impl.cu (hunk 13 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/cuda/math/topk_impl.cu]: Done.

File: onnxruntime/core/providers/cuda/math/topk_impl.cu (hunk 14 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/cuda/math/topk_impl.cu]: Done.

File: onnxruntime/core/providers/cuda/math/topk_impl.cu [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+    output_i[id] = dimension;
+  }
+}
+
+template <typename T>
+Status TopKImpl(const CudaKernel* kernel, const T* input_x, T* output_v, int64_t* output_i, const int64_t* elem_nums, size_t size, int64_t axis, int64_t K, int64_t largest, int64_t sorted, int64_t N, int64_t dimension) {
+  CudaKernel::CudaAsyncBuffer<T> input_key_buffer(kernel, dimension);
+  CudaKernel::CudaAsyncBuffer<T> output_key_buffer(kernel, dimension);
+  CudaKernel::CudaAsyncBuffer<int64_t> input_value_buffer(kernel, dimension);
+  CudaKernel::CudaAsyncBuffer<int64_t> output_value_buffer(kernel, dimension);
```
Comment (yufenglee) [inline onnxruntime/core/providers/cuda/math/topk_impl.cu]: You can use GetScratchBuffer as you don't need to copy data from cpu to gpu

Comment (RandySheriffH) [reply onnxruntime/core/providers/cuda/math/topk_impl.cu]: Done.
